app/repositories/invoice_repository.rb

```diff
@@ -0,0 +1,26 @@
+class InvoiceRepository
```

Comment:
Contrary to the product repository that encapsulates a composite data model, the invoice repository just forwards operations to the model.

The tradeoff I see is that it creates more overhead but if one day the nature of data model of invoice changes (several tables, external service), we won't have to update everywhere it's used.

How do you see the balance in the tradeoff, do you think that whenever we create a model we should hide it behind a repository?

Reply:
Indeed, I would create a Repository for each entity I have. It's because the repository offer a business interface that should match business needs and restrict available operations to control them.
At first, it may look like an overhead as we are just forwarding basic operations but it already restrict the power (limited operations instead full power) and can be injected and mocked easily. Later, with more complexity, business rules, validations or complex storage, it would be necessary.